proxy/sources/mangadventure.py

```diff
@@ -0,0 +1,156 @@
+from datetime import datetime as dt
+from json import loads
+from urllib.parse import urlparse
+
+from django.http.response import HttpResponseBadRequest
+from django.shortcuts import redirect
+from django.urls.conf import re_path
+
+from ..source import ProxySource
+from ..source.data import ChapterAPI, SeriesAPI, SeriesPage
+from ..source.helpers import api_cache, get_wrapper, encode, decode
+
+
+class MangAdventure(ProxySource):
+    headers = {"Referer": "https://cubari.moe/"}
+
+    whitelist = (
+        # "127.0.0.1",
+        "arc-relight.com",
+        "www.arc-relight.com",
+        "assortedscans.com",
+        "helveticascans.com",
+        "mangadventure.herokuapp.com",
+    )
+
+    def get_reader_prefix(self) -> str:
+        return "mangadventure"
+
+    def shortcut_instantiator(self) -> list:
+        def handler(_, raw_url: str):
+            url = urlparse(raw_url)
+            if url.hostname not in self.whitelist:
+                return HttpResponseBadRequest()
+            if not url.path.startswith("/reader/"):
+                return HttpResponseBadRequest()
+            base = f"{url.scheme}/{url.netloc}/"
+            path = url.path[8:].rstrip("/").split("/")
+            size = len(path)
+            if size == 1:
+                return redirect(
+                    f"reader-{self.get_reader_prefix()}-series-page",
+                    encode(base + path[0])
+                )
+            if size == 3:
+                vol, num = path[1], path[2].replace(".", "-")
+                return redirect(
+                    f"reader-{self.get_reader_prefix()}-chapter-page",
+                    encode(f"{base}{path[0]}/{vol}/{num}"), num, "1"
+                )
+            if size == 4:
+                vol, num = path[1], path[2].replace(".", "-")
+                return redirect(
+                    f"reader-{self.get_reader_prefix()}-chapter-page",
+                    encode(f"{base}{path[0]}/{vol}/{num}"), num, path[3]
+                )
```

Comment:
It does.

Reply:
I think this may break depending on your entrypoint. The slug will only be read by `series_api_handler` and `series_page_handler`, since the `chapter_api_handler` only handles slugs generated by the code you have on line 91. It should be sufficient to pass in `encode(base + path[0])` for all these redirects.

I'll merge this after that change.

Reply:
There were no issues when I tested it. Do you have an idea of what might break it?

Reply:
I tested these two URLs:

```
{cubari_instance}/ma/https://assortedscans.com/reader/cuprums_bride/
```

and

```
{cubari_instance}/ma/https://assortedscans.com/reader/cuprums_bride/2/14/1/
```

The latter encodes the volume and chapter in the slug, which causes the external resource to return a 501.

Reply:
This will now redirect everything to the series page.

Reply:
I don't think it would work properly without the volume number.

If the proxy is still relevant when we rewrite the frontend, I might make it use the chapter ID.

Reply:
The redirect would use whatever chapter number is assigned [here](https://github.com/subject-f/cubarimoe/pull/7/files#diff-03960ca97dfe5e46132bdb0087b70375df486632e230701448adcc9ca144da44R68-R84), so if a chapter is identified as a composite of a volume and chapter (ie. vol1ch1, vol2ch1), it'd break Cubari's chapter invariant regardless. Otherwise, the chapter's contents are resolved from the series API response.

If that's still fine with you then I can merge.

Reply:
Merge it. ¯\\\_(ツ)_/¯